libcaf_net/test/doorman.cpp

```diff
@@ -0,0 +1,161 @@
+    mpx->set_thread_id();
+    CAF_CHECK_EQUAL(mpx->num_socket_managers(), 1u);
+    auth.port = 0;
+    auth.host = "0.0.0.0"s;
+  }
+
+  bool handle_io_event() override {
+    return mpx->poll_once(false);
+  }
+
+  multiplexer_ptr mpx;
+  uri::authority_type auth;
+};
+
+class dummy_application {
+public:
+  static expected<std::vector<byte>> serialize(actor_system& sys,
+                                               const type_erased_tuple& x) {
+    std::vector<byte> result;
+    binary_serializer sink{sys, result};
+    if (auto err = message::save(sink, x))
+      return err.value();
+    return result;
+  }
+
+  template <class Parent>
+  error init(Parent&) {
+    return none;
+  }
+
+  template <class Parent>
+  void write_message(Parent& parent,
+                     std::unique_ptr<endpoint_manager_queue::message> msg) {
+    parent.write_packet(msg->payload);
+  }
+
+  template <class Parent>
+  error handle_data(Parent&, span<const byte>) {
+    return none;
+  }
+
+  template <class Parent>
+  void resolve(Parent&, string_view path, const actor& listener) {
+    anon_send(listener, resolve_atom::value,
+              "the resolved path is still "
+                + std::string(path.begin(), path.end()));
+  }
+
+  template <class Parent>
+  void timeout(Parent&, atom_value, uint64_t) {
+    // nop
+  }
+
+  template <class Parent>
+  void new_proxy(Parent&, actor_id) {
+    // nop
+  }
+
+  template <class Parent>
+  void local_actor_down(Parent&, actor_id, error) {
+    // nop
+  }
+
+  void handle_error(sec) {
+    // nop
+  }
+};
+
+class dummy_application_factory {
+public:
+  using application_type = dummy_application;
+
+  static expected<std::vector<byte>> serialize(actor_system& sys,
+                                               const type_erased_tuple& x) {
+    return dummy_application::serialize(sys, x);
+  }
+
+  template <class Parent>
+  error init(Parent&) {
+    return none;
+  }
+
+  application_type make() const {
+    return dummy_application{};
+  }
+};
+
+} // namespace
+
+CAF_TEST_FIXTURE_SCOPE(doorman_tests, fixture)
+
+CAF_TEST(doorman accept) {
+  auto acceptor = unbox(make_tcp_accept_socket(auth, false));
+  auto port = unbox(local_port(socket_cast<network_socket>(acceptor)));
+  auto acceptor_guard = make_socket_guard(acceptor);
+  CAF_MESSAGE("opened acceptor on port " << port);
+  auto mgr = make_endpoint_manager(
+    mpx, sys,
+    doorman<dummy_application_factory>{acceptor_guard.release(),
+                                       dummy_application_factory{}});
+  CAF_CHECK_EQUAL(mgr->init(), none);
+  auto before = mpx->num_socket_managers();
+  CAF_CHECK_EQUAL(before, 2u);
+  uri::authority_type dst;
+  dst.port = port;
+  dst.host = "localhost"s;
+  CAF_MESSAGE("connecting to doorman on: " << dst);
+  auto conn = make_socket_guard(unbox(make_connected_tcp_stream_socket(dst)));
+  CAF_MESSAGE("waiting for connection");
+  while (mpx->num_socket_managers() != before + 1)
```

Comment:
This needs an additional exit condition, because otherwise the loop can run forever. I go ahead and merge, but this needs a followup.

Reply:
Will provide that tomorrow! :+1: